### PR TITLE
Cdf 8785 pnid detect entities 

### PR DESCRIPTION
**Pull request by chingyu-cognite**

### CODE REVIEW

File: cognite/experimental/_api/pnid_parsing.py [hunk 1]
```diff
@@ -19,13 +20,16 @@ def detect(
 
         Args:
             file_id (int): ID of the file, should already be uploaded in the same tenant.
-            entities (List[str]): List of entities to detect
+            entities (List[Union[str, dict]]): List of entities to detect
+            search_field (str): the field to search in entities, only relevant if entities is a list of dict
```
Comment (rebwib) [inline cognite/experimental/_api/pnid_parsing.py:24]:
```suggestion
            search_field (str): If entities is a list of dictionaries, this is the field to detect in the PnId 
```

Comment (rebwib) [reply cognite/experimental/_api/pnid_parsing.py:24]: Or something?

File: cognite/experimental/_api/pnid_parsing.py [hunk 2]
```diff
@@ -19,13 +20,16 @@ def detect(
...
             name_mapping (Dict[str,str]): Optional mapping between entity names and their synonyms in the P&ID. Used if the P&ID contains names on a different form than the entity list (e.g a substring only). The response will contain names as given in the entity list.
             partial_match (bool): Allow for a partial match (e.g. missing prefix).
             min_tokens (int): Minimal number of tokens a match must be based on
         Returns:
             ContextualizationJob: Resulting queued job. Note that .results property of this job will block waiting for results."""
-        return self._run_job(
+        entities, entities_return = self._detect_before_hook(entities, search_field)
```
Comment (rebwib) [inline cognite/experimental/_api/pnid_parsing.py:30]: Should we add a test that all elements in entities are either strings or objects?

Comment (chingyu-cognite) [reply cognite/experimental/_api/pnid_parsing.py:33]: yes, added

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +12,10 @@ Changes are grouped as follows
 - `Fixed` for any bug fixes.
 - `Security` in case of vulnerabilities.
 
+## [0.22.1] - 2020-09-09
+### Added
+- Added `/detect` in Pnid Parser can accept entities with type `List[Union[str, dict]]` and return `entities` if the type of input `entities` has type of `List[dict]`
```
Comment (rebwib) [inline CHANGELOG.md:17]:
```suggestion
- `/detect` in Pnid Parser can accept entities with type `List[Union[str, dict]]` and returns `entities` if the type of input `entities` has type of `List[dict]`
```

File: cognite/experimental/_api/pnid_parsing.py [hunk 4]
```diff
@@ -19,13 +20,19 @@ def detect(
...
-            entities (List[str]): List of entities to detect
+            entities (List[Union[str, dict]]): List of entities to detect
+            search_field (str): If entities is a list of dictionaries, this is the key to the values to detect in the PnId 
             name_mapping (Dict[str,str]): Optional mapping between entity names and their synonyms in the P&ID. Used if the P&ID contains names on a different form than the entity list (e.g a substring only). The response will contain names as given in the entity list.
             partial_match (bool): Allow for a partial match (e.g. missing prefix).
             min_tokens (int): Minimal number of tokens a match must be based on
         Returns:
             ContextualizationJob: Resulting queued job. Note that .results property of this job will block waiting for results."""
-        return self._run_job(
+        assert all([isinstance(entity, str) for entity in entities]) or all(
```
[the remark this answers is not in the document]
Comment (chingyu-cognite) [reply cognite/experimental/_api/pnid_parsing.py:30]: what do you mean? I though it raises an `AssertionError`

File: cognite/experimental/_api/pnid_parsing.py [hunk 5]
```diff
@@ -34,6 +41,31 @@ def detect(
             name_mapping=name_mapping,
             min_tokens=min_tokens,
         )
+        job.result
```
Comment (rebwib) [inline cognite/experimental/_api/pnid_parsing.py:44]: Used for debugging?

File: cognite/experimental/_api/pnid_parsing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rebwib) [reply cognite/experimental/_api/pnid_parsing.py:30]: Maybe add a `if all([isinstance(entity, str) for entity in entities]) ` then entities_return = entities.copy(), elif `if all([isinstance(entity, dict) for entity in entities]) `, do the convert, else: raise error?

File: cognite/experimental/_api/pnid_parsing.py (hunk 5 above)
Comment (chingyu-cognite) [reply cognite/experimental/_api/pnid_parsing.py:44]: no, `job.result ` will make sure the job compeleted ( or failed)

File: cognite/experimental/_api/pnid_parsing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rebwib) [reply cognite/experimental/_api/pnid_parsing.py:30]: > what do you mean? I though it raises an `AssertionError`

In other folders we have used `assert` in tests and `raise` for instance `ValueError` if there is something wrong with the input

File: cognite/experimental/_api/pnid_parsing.py (hunk 5 above)
Comment (rebwib) [reply cognite/experimental/_api/pnid_parsing.py:44]: Ok, we use the `wait_for_completion` method for this, https://github.com/cognitedata/cognite-sdk-python-experimental/blob/1c0141acae4d7976bfb79f0d466093162e59fb04/cognite/experimental/data_classes/contextualization.py#L134

See for instance predict for entity matcher, https://github.com/cognitedata/cognite-sdk-python-experimental/blob/1c0141acae4d7976bfb79f0d466093162e59fb04/cognite/experimental/data_classes/contextualization.py#L162

File: cognite/experimental/_api/pnid_parsing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chingyu-cognite) [reply cognite/experimental/_api/pnid_parsing.py:30]: Why change the name to something with `convert`  doesnt it make it confusing with the endpoint `convert`?

[the remark this answers is not in the document]
Comment (chingyu-cognite) [reply cognite/experimental/_api/pnid_parsing.py:30]: The reason I used `xx_{before,after}_hook` is to signify this is something we do right before(after) the main function(xx) is called. I feel it's modular this way.

[the remark this answers is not in the document]
Comment (chingyu-cognite) [reply cognite/experimental/_api/pnid_parsing.py:30]: > > what do you mean? I though it raises an `AssertionError`
> 
> In other folders we have used `assert` in tests and `raise` for instance `ValueError` if there is something wrong with the input

I can do this.

[the remark this answers is not in the document]
Comment (rebwib) [reply cognite/experimental/_api/pnid_parsing.py]: > Why change the name to something with `convert` doesnt it make it confusing with the endpoint `convert`?

Sure, mostly ment as a suggestion. No need to rename.

**Review by rebwib — APPROVED** [review]
LGTM, @sanderland do you want to have a look as well?

**Pull request merged**